### PR TITLE
keep properties of invalid schemas

### PR DESCRIPTION
**Pull request by sberan**
**What issue does this pull request resolve?**

https://github.com/epoberezkin/ajv/issues/129#issuecomment-325155599

**What changes did you make?**

Only remove additional properties when the schema is valid.

### CODE REVIEW

**Review by epoberezkin — CHANGES_REQUESTED** [review]
Thank you for the PR!

1. The change of behaviour is only possible for a specific value of removeAdditional option, as you proposed in https://github.com/epoberezkin/ajv/issues/129#issuecomment-325134083, it cannot be done for all option values. Not sure if "afterValidation" is clear enough, I will think about it, but the main point is that it should be something specific. Current values should keep the current behaviour.
2. In addition to the test you added (where the option value should be changed), there should be two more tests 
    - for the option `removeAdditional: true` the property is removed even though the validation failed (the current behaviour)
    - for the new option value the property is removed when validation passed

An open question is how would be handled the situation when additionalProperties is a schema? There three possibilities
1. just preserve old behaviour
2. we need two new option values (new behaviour for current options `false` and `"fail"`)
3. create a completely new option, e.g. `removeAdditionalOnSuccess`: true

I need to sleep on it, but the last choice seems better.
For choices 2 or 3 there will be needed additional 3 tests (same as above) for cases when additionalProperties is a schema.

What do you think?

File: spec/options.spec.js [hunk 1]
```diff
@@ -43,43 +52,39 @@ describe('Ajv Options', function () {
...
+        fizz: 'buzz'
       };
 
-      ajv.validate('//test/fooBar2', object).should.equal(true);
-      object.should.have.property('foo');
-      object.should.have.property('bar');
-      object.should.have.property('baz');
-      object.should.not.have.property('fizz');
+      console.log(ajv.getSchema('//test/fooBar2').toString());
+      ajv.validate('//test/fooBar2', failingObject).should.equal(false);
```
Comment (sberan) [inline spec/options.spec.js]: @epoberezkin I think I found a bug in master. When `additionalProperties` is a schema, properties are not validated (and this test fails at this location 😬 .

Comment (sberan) [reply spec/options.spec.js]: (it's only a problem if you supply an invalid additional property with an additional properties schema, and `removeAdditional: 'failing'` )

Comment (sberan) [reply spec/options.spec.js]: @epoberezkin I figured out the issue. `$nextValid` was not being unset when the additional property was removed, which caused the function to exit with no errors.

**Comment by sberan** [comment]
@epoberezkin I can go with `removeAdditionalOnSuccess` - I've implemented it in this PR and added tests for all the cases noted. What do you think?

**Comment by sberan** [comment]
@epoberezkin note that the diff is much easier to read if you ignore whitespace (`?w=1`) 😄

**Comment by epoberezkin** [comment]
@sberan I will review, thank you

File: lib/dot/properties.jst [hunk 2]
```diff
@@ -121,7 +131,8 @@ var {{=$nextValid}} = true;
               if (errors) validate.errors.length = errors;
               else validate.errors = null;
             }
-            delete {{=$data}}[{{=$key}}];
+            {{=$nextValid}} = true;
```
Comment (sberan) [inline lib/dot/properties.jst:134]: setting this field to true fixed the bug I mentioned.

**Comment by epoberezkin** [comment]
@sberan Thank you!

I don't think it is going to work as expected (that is to only remove additional properties when otherwise current schema succeeds). In your code additional properties are removed when all other properties succeed, but it doesn't check that other keywords succeed, e.g. `required` or `minProperties` etc.

Also the changes in tests are difficult to follow. I suggest not touching the existing tests and instead add other tests that validate that without this new option even if schema fails additional props are removed anyway and with this option they don't.

Regarding the fix for 'failing' option - I still don't understand the issue. Maybe it's better to make it a separate PR or at least make a separate test case that demonstrates the issue.

**Comment by epoberezkin** [comment]
Closing, it is old.

**Pull request closed without merging**